### PR TITLE
feat(NODE-6503): Add Int32 as a SchemaType

### PR DESCRIPTION
**Pull request by baileympearson**
Note: @aditi-khare-mongoDB is the implementer, just a clarification in case it seems confusing why @baileympearson is leaving review comments.

### Summary

**Motivation**
Support a 32-bit integer type in mongoose for compatibility with CSFLE/QE. The current `Number` schema type can result in either Int32, Int64, or Doubles that are inserted into the database using type inference but CSFLE and QE require exact BSON types in schemas.

**Summary of Changes**
- Add `SchemaType` `Int32`, which has the following behavior:
  - its default `cast` method throws when provided a 
    - a non-numeric string
    - a non-integer value
    - a `NaN`
    - a number beyond the bounds of INT32_MAX or INT32_MIN
  - it is queryable as `$type: int` or `$type: number`, but NOT `$type: double`
 - Add API Docs for new query option and new schema type

### Example

```
const vehicleSchema = new Schema({ numberOfTires: mongoose.Int32 });
```

### CODE REVIEW

File: lib/schema.js [hunk 1]
```diff
@@ -2848,6 +2848,7 @@ module.exports = exports = Schema;
  * - [Mixed](https://mongoosejs.com/docs/schematypes.html#mixed)
  * - [UUID](https://mongoosejs.com/docs/schematypes.html#uuid)
  * - [BigInt](https://mongoosejs.com/docs/schematypes.html#bigint)
+ * - Int32
```
Comment (aditi-khare-mongoDB) [inline lib/schema.js]: Same here - once documentation is created for the Int32 schema type, this should be updated with a link - for now I can also include a BSON link

File: lib/browser.js [hunk 2]
```diff
@@ -50,6 +50,7 @@ exports.Schema = require('./schema');
  * - [ObjectId](https://mongoosejs.com/docs/schematypes.html#objectids)
  * - [Map](https://mongoosejs.com/docs/schematypes.html#maps)
  * - [Subdocument](https://mongoosejs.com/docs/schematypes.html#schemas)
+ * - Int32
```
Comment (aditi-khare-mongoDB) [inline lib/browser.js]: Once documentation is created for the Int32 schema type, this should be updated with a link - for now I can also include a BSON link.

File: lib/schema/int32.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
+'use strict';
```
Comment (aditi-khare-mongoDB) [inline lib/schema/int32.js:1]: This might seem like a lot of changes but the skeleton is the same from the other numeric SchemaType files`schema/bigint.js` and `schema/decimal128.js` files.

File: docs/schematypes.md [hunk 4]
```diff
@@ -647,6 +649,21 @@ const question = new Question({ answer: 42n });
...
+Int32s are stored as [32-bit integers in MongoDB (BSON type "int")](https://www.mongodb.com/docs/manual/reference/bson-types/).
+
+```javascript
+const studentsSchema = new Schema({
+  id: Int32
+});
+const Student = mongoose.model('Student', schema);
+
+const student = new Student({ id: 1339 });
+typeof student.id; // 'number'
```
Comment (baileympearson) [inline docs/schematypes.md:664]: Can we note that Int32s throw if the value is not coercible losslessly to an int32?

File: lib/cast/int32.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  if (val === '') {
+    return null;
+  }
+
+  if (typeof val === 'string' || typeof val === 'number') {
+
+    const INT32_MAX = 0x7FFFFFFF;
+    const INT32_MIN = -0x80000000;
+
+    const _val = Number(val);
```
Comment (baileympearson) [inline lib/cast/int32.js]:
```suggestion
    const coercedValue = Number(val);
```
perhaps a slightly better name here?

File: lib/schema/int32.js [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+  if (arguments.length === 0) {
+    return this._cast;
+  }
+  if (caster === false) {
+    caster = this._defaultCaster;
+  }
+  this._cast = caster;
+
+  return this._cast;
+};
```
Comment (baileympearson) [inline lib/schema/int32.js]: We should probably add coverage for getting / setting custom cast functions, if we're going to support it.  `Int32.cast(false)` disables casting of numeric values - we should test that (we likely also need to define _defaultCaster, so that if somebody uses it we don't throw an error because you can't call undefined as a function).

File: lib/schema/int32.js [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+/*!
+ * ignore
+ */
+
+SchemaInt32._checkRequired = v => v != null && isBsonType(v, 'Int32');
```
Comment (baileympearson) [inline lib/schema/int32.js]:
```suggestion
SchemaInt32._checkRequired = v => v != null
```

If, when we cast the value as `Int32`, we don't return an instance of our `Int32` class, we shouldn't check that it is a BSON type class.   (see my comment about `required` in the test file)

Can we also add a test that when creating a required int32 in a schema, we throw a validation error if it isn't present or coercible to int32?

File: lib/cast/int32.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+module.exports = function castInt32(val) {
+  if (val == null) {
+    return val;
+  }
+  if (val === '') {
+    return null;
+  }
+
```
Comment (baileympearson) [inline lib/cast/int32.js:23]: We're missing the case where we receive an Int32 typeclass:

```
castInt32(new Int32(3)) // 3
```

File: lib/schema/int32.js [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+ * Casts to Int32
+ *
+ * @param {Object} value
+ * @param {Object} model this value is optional
+ * @api private
+ */
+
+SchemaInt32.prototype.cast = function(value) {
+  let castInt32;
+  if (isBsonType(value, 'Int32')) {
```
Comment (baileympearson) [inline lib/schema/int32.js]: I think this should be moved into the cast function, but also, when we do receive a BSON Int32 class we'd want to return the value, not the class itself (I mention this in cast/int32.js)

File: lib/schema/int32.js [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+    throw err;
+  }
+};
+
+/**
+ *
+ * @api private
+ */
+
+SchemaInt32.prototype._castNullish = function _castNullish(v) {
```
Comment (baileympearson) [inline lib/schema/int32.js]: Do you know what the _castNullish function is for?  I don't see it implemented for every BSON type :thinking

File: lib/schema/int32.js [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+  } catch (error) {
+    throw new CastError('Int32', value, this.path, error, this);
+  }
+};
+
+/*!
+ * ignore
+ */
+
+SchemaInt32.$conditionalHandlers = {
```
Comment (baileympearson) [inline lib/schema/int32.js:198]: I assume this is just copied from the plugin?

Comment (baileympearson) [reply lib/schema/int32.js:198]: Actually it seems like the plugin doesn't override anything: https://github.com/vkarpov15/mongoose-int32/blob/26a6c76cc56c24173de13e8a0f3c6401ce867c9b/int32.js#L41

Do you know what `$conditionalHandlers` is used for?

File: lib/types/index.js [hunk 12]
```diff
@@ -14,6 +14,7 @@ exports.Embedded = require('./arraySubdocument');
 exports.DocumentArray = require('./documentArray');
 exports.Decimal128 = require('./decimal128');
 exports.ObjectId = require('./objectid');
+exports.Int32 = require('./int32');
```
Comment (baileympearson) [inline lib/types/index.js]: Unlike other schema types here, we don't need a type (I don't think).  The types are mongoose-specific data types.  Notice that not all schema types have associated mongoose types (String or boolean, for example).  We can remove everything related to a mongoose Type from the PR.

File: lib/utils.js [hunk 13]
```diff
@@ -93,6 +93,10 @@ exports.deepEqual = function deepEqual(a, b) {
     return a.toString() === b.toString();
   }
 
+  if (isBsonType(a, 'Int32') && isBsonType(b, 'In32')) {
+    return a.value === b.value();
+  }
```
Comment (baileympearson) [inline lib/utils.js]: We don't need this change, deepEquality already worked for Int32 classes.  But if we do leave it in this PR:

```suggestion
  if (isBsonType(a, 'Int32') && isBsonType(b, 'Int32')) {
    return a.value === b.value;
  }
```

File: test/int32.test.js [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+    const Test = mongoose.model('Test', schema);
+
+    const doc = new Test({
+      myInt32: 13
+    });
+    assert.strictEqual(doc.myInt32, 13);
+    assert.equal(typeof doc.myInt32, 'number');
+  });
+
+  it('supports the \'required\' property', async function() {
```
Comment (baileympearson) [inline test/int32.test.js]: (entirely optional)

Usually we'd just use a different outer quote:
```suggestion
  it(`supports the 'required' property`, async function() {
```

File: test/int32.test.js [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+    const schema = new Schema({
+      int32: {
+        type: Schema.Types.Int32,
+        required: true
+      }
+    });
+    const Test = mongoose.model('Test', schema);
+
+    const doc = new Test({
```
Comment (baileympearson) [inline test/int32.test.js]: It might be worth adding a test for a non-null required value as well, simply because local testing of your PR caught that:

```typescript
new Test({ 
	int: 3
})
```

still throws, because your `checkRequired` function is expecting an instance of our BSON Int32 class

File: test/int32.test.js [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+    it('casts from number', function() {
+      const schema = new Schema({
+        int2: Schema.Types.Int32
+      });
+      const Test = mongoose.model('Test', schema);
+
+      const doc = new Test({
+        int2: '-997.0'
+      });
+      assert.strictEqual(doc.int2, -997);
```
Comment (baileympearson) [inline test/int32.test.js]: Can we add casts from BitInt, Int32, maybe Long as well?

File: test/int32.test.js [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+    let Test;
+
+    beforeEach(function() {
+      const schema = new Schema({
+        myInt: Schema.Types.Int32
+      });
+      Test = mongoose.model('Test', schema);
+    });
+
+    describe('when a non-integer input is provided to an Int32 field', () => {
```
Comment (baileympearson) [inline test/int32.test.js]:
```suggestion
    describe('when a decimal is provided to an Int32 field', () => {
```

File: test/int32.test.js [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+        assert.ok(err.errors['myInt']);
+        assert.equal(err.errors['myInt'].name, 'CastError');
+        assert.equal(
+          err.errors['myInt'].message,
+          'Cast to Int32 failed for value "-42.4" (type number) at path "myInt"'
+        );
+      });
+    });
+
+    describe('when a non-numeric string is provided to an Int32 field', () => {
```
Comment (baileympearson) [inline test/int32.test.js:311]: Can we add a test for a numeric decimal string?

File: test/int32.test.js [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+      });
+
+      it('is queryable as a BSON Int32 in MongoDB', async function() {
+        await Test.create({ myInt: '42' });
+        const doc = await Test.findOne({ myInt: { $type: 'int' } });
+        assert.ok(doc);
+        assert.strictEqual(doc.myInt, 42);
+      });
+
+      it('is queryable as a BSON Double in MongoDB', async function() {
```
Comment (baileympearson) [inline test/int32.test.js]: Doesn't this test imply that it is not queryable as a double?

File: docs/schematypes.md (hunk 4 above)
Comment (aditi-khare-mongoDB) [reply docs/schematypes.md:664]: Yep, done! And added some more tests for the cases in which a `CastError` should not be thrown.

File: test/int32.test.js (hunk 16 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js]: Yup! Just added them!

File: lib/cast/int32.js (hunk 8 above)
Comment (aditi-khare-mongoDB) [reply lib/cast/int32.js:23]: Added it in!

File: lib/cast/int32.js (hunk 5 above)
Comment (aditi-khare-mongoDB) [reply lib/cast/int32.js]: Made the change!

File: lib/schema/int32.js (hunk 6 above)
Comment (aditi-khare-mongoDB) [reply lib/schema/int32.js]: Yep! Just added!

File: lib/schema/int32.js (hunk 7 above)
Comment (aditi-khare-mongoDB) [reply lib/schema/int32.js]: Removed the unnecessary check and added the test!

File: test/int32.test.js (hunk 18 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js:311]: Yep added it in!

File: test/int32.test.js (hunk 19 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js]: That's a typo, my bad!

File: test/int32.test.js (hunk 17 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js]: I'm going to put 'non-integer decimal'  here since, 'decimals like `1.0` are still valid inputs.

File: lib/utils.js (hunk 13 above)
Comment (aditi-khare-mongoDB) [reply lib/utils.js]: I removed the change, good catch!

File: lib/types/index.js (hunk 12 above)
Comment (aditi-khare-mongoDB) [reply lib/types/index.js]: Are decimal128 and objectId mongoose specific data types?

File: lib/schema/int32.js (hunk 11 above)
Comment (aditi-khare-mongoDB) [reply lib/schema/int32.js:198]: I believe it handles casting for queries? I'm not sure exactly why its needed but all the other types have `$conditionalHandlers` defined, so I figured its best to keep it for consistency.

File: lib/schema/int32.js (hunk 10 above)
Comment (aditi-khare-mongoDB) [reply lib/schema/int32.js]: We don't need it for Int32s, since null is a valid input. I removed this piece of code, good catch.

File: test/int32.test.js (hunk 15 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js]: Done!

File: docs/schematypes.md [hunk 20]
```diff
@@ -647,6 +649,44 @@ const question = new Question({ answer: 42n });
...
+call it and assign the returned value to the path.
+
+The values `null` and `undefined` are not cast.
+
+The following inputs will result will all result in a [CastError](validation.html#cast-errors) once validated, meaning that it will not throw on initialization, only when validated:
+- NaN
+- strings that cast to NaN
+- objects that don't have a `valueOf()` function
+- a decimal that must be rounded to be an integer
+- an input that represents a value beyond the bounds of an 32-bit integer
```
Comment (baileympearson) [inline docs/schematypes.md]:
```suggestion
- an input that represents a value ouside the bounds of an 32-bit integer
```

File: lib/schema/int32.js [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+/*!
+ * ignore
+ */
+
+SchemaInt32._defaultCaster = v => {
+  const INT32_MAX = 0x7FFFFFFF;
+  const INT32_MIN = -0x80000000;
+
+  if (v != null) {
+    if (typeof v !== 'number') {
```
Comment (baileympearson) [inline lib/schema/int32.js]: do you think we should also confirm that it is an integer here?

File: lib/types/index.js (hunk 12 above)
Comment (baileympearson) [reply lib/types/index.js]: No - but mongoose has specific wrappers for them in lib/types.  see https://github.com/Automattic/mongoose/issues/1671

File: test/int32.test.js [hunk 22]
```diff
@@ -174,6 +386,53 @@ describe('Int32', function() {
     });
   });
 
+  describe('custom casters', () => {
+    const defaultCast = mongoose.Schema.Types.Int32.cast();
+
+    afterEach(() => {
+      mongoose.Schema.Types.Int32.cast(defaultCast);
+    });
```
Comment (baileympearson) [inline test/int32.test.js:407]:
```suggestion
  	afterEach(() => {
      mongoose.Schema.Types.Int32.cast(false);
    });
```

File: test/int32.test.js [hunk 23]
```diff
@@ -174,6 +386,53 @@ describe('Int32', function() {
...
+        myInt2: {
+          type: Schema.Types.Int32
+        }
+      });
+      const Test = mongoose.model('Test', schema);
+      const doc = new Test({
+        myInt: '52',
+        myInt2: 52
+      });
+      assert.strictEqual(doc.myInt1, undefined);
```
Comment (baileympearson) [inline test/int32.test.js]: How is this test passing? from the logic in the default caster, i'd expect it to throw an error because `typeof <string>` is not equal to `number`

**Review by baileympearson — COMMENTED** [review]
besides removing the code from `lib/types`, only minor comments.  Lint is failing too, btw

(also can't request changes)

File: test/int32.test.js (hunk 23 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js]: By default Schema setters will set the value to undefined and set an error, I can add a check for the error, too.

File: test/int32.test.js (hunk 22 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js:407]: Your suggested change means we're turning casting off (setting to `_defaultCast`), but we want to go back to the initial behavior which is `defaultCast` variable here (cast/int32.js's `cast` function). I suggest we keep the current code.

File: lib/schema/int32.js (hunk 21 above)
Comment (aditi-khare-mongoDB) [reply lib/schema/int32.js]: I'm not sure - I wonder if that would make it too similar to the other cast function

Comment (aditi-khare-mongoDB) [reply lib/schema/int32.js]: We can leave a comment open for when we open this for Val

File: lib/types/index.js (hunk 12 above)
Comment (aditi-khare-mongoDB) [reply lib/types/index.js]: I see, I removed int32 from types.d.ts!

File: docs/schematypes.md (hunk 20 above)
Comment (aditi-khare-mongoDB) [reply docs/schematypes.md]: Done!

File: types/inferschematype.d.ts [hunk 24]
```diff
@@ -219,27 +219,31 @@ type IsSchemaTypeFromBuiltinClass<T> = T extends (typeof String)
             ? true
             : T extends (typeof Schema.Types.Decimal128)
               ? true
```
Comment (baileympearson) [inline types/inferschematype.d.ts:221]: it seems like there are a lot of unintended changes in this file?

File: lib/schema/int32.js (hunk 11 above)
Comment (baileympearson) [reply lib/schema/int32.js:198]: I agree that we need conditional handlers, but we also need to understand if we should be overriding any of the default $conditionalHandlers.  

My hunch is that we should be leaving them alone, because Val's plugin does not override the defaults: vkarpov15/mongoose-int32@26a6c76/int32.js#L41

Comment (aditi-khare-mongoDB) [reply lib/schema/int32.js:198]: If we leave them alone / don't define them in this file then the conditional operators schemaType tests fail; I suggest we keep them in here. Val's plug-in was missing a few important parts of the implementation.

File: types/inferschematype.d.ts (hunk 24 above)
Comment (aditi-khare-mongoDB) [reply types/inferschematype.d.ts:221]: Removed some of the changes, no longer necessary since mongoose.Type.Int32 doesn't exist. The diff looks larger than it is due to the indentation its just a one line change to add `Schema.Types.Int32`

File: docs/schematypes.md [hunk 25]
```diff
@@ -647,6 +649,44 @@ const question = new Question({ answer: 42n });
...
+const studentsSchema = new Schema({
+  id: Int32
+});
+const Student = mongoose.model('Student', schema);
+
+const student = new Student({ id: 1339 });
+typeof student.id; // 'number'
+```
+
+There are several types of values that will be successfully cast to a Number.
```
Comment (baileympearson) [inline docs/schematypes.md:667]:
```suggestion
There are several types of values that will be successfully cast to an Int32.
```

File: lib/mongoose.js [hunk 26]
```diff
@@ -1138,6 +1139,22 @@ Mongoose.prototype.syncIndexes = function(options) {
...
+ * Do not use this to create a new Int32 instance, use `mongoose.Types.Int32`
+ * instead.
+ *
+ * #### Example:
+ *
+ *     const vehicleSchema = new Schema({ numTires: mongoose.Int32 });
+ *
+ * @property Int32
+ * @api public
+ */
```
Comment (baileympearson) [inline lib/mongoose.js]: Do we need to store a reference to our schema on the Mongoose object?  My thought is no becase the majority of schema types do not.  Can we remove this?  (and the TS changes you made as well)?

File: index.js [hunk 27]
```diff
@@ -46,6 +46,7 @@ module.exports.Decimal128 = mongoose.Decimal128;
 module.exports.Mixed = mongoose.Mixed;
 module.exports.Date = mongoose.Date;
 module.exports.Number = mongoose.Number;
+module.exports.Int32 = mongoose.Int32;
```
Comment (baileympearson) [inline index.js]: Along with my comment in mongoose.js, i think this should be removed

**Review by baileympearson — COMMENTED** [review]
just one minor comment

File: lib/cast/int32.js [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+  let coercedVal;
+  if (val instanceof BSON.Int32 || val instanceof BSON.Double) {
+    coercedVal = val.value;
+  } else if (val instanceof BSON.Long) {
+    coercedVal = val.toNumber();
+  } else {
+    coercedVal = Number(val);
+  }
```
Comment (nbbeeken) [inline lib/cast/int32.js]: I noticed the checks for Int32 and Double types here, and I was curious about the reasoning behind handling them individually versus using Number(val) across the board. The type handling makes sense if there’s a specific case where it’s essential to retain `toNumber()` for Long, as it helps avoid unnecessary overhead, particularly with `Long` values (where a `toString` conversion could be expensive, or `Int32`/`Double` can be cast with just a property access rather than a function call). Could we add a unit test that would flag if this handling were altered? That way, future updates won't accidentally bypass any valuable optimizations here.

(edit: tone)

File: lib/cast/int32.js [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+module.exports = function castInt32(val) {
+  if (val == null) {
+    return val;
+  }
+  if (val === '') {
+    return null;
+  }
+
+  let coercedVal;
+  if (val instanceof BSON.Int32 || val instanceof BSON.Double) {
```
Comment (nbbeeken) [inline lib/cast/int32.js]: we check for `_bsontype` because that lines up with how the serializer interprets our BSON types. `instanceof` is a different contract than our code expects so I think it shouldn't be introduced here so we don't create two versions of what is a true "type check"

File: index.js (hunk 27 above)
Comment (vkarpov15) [reply index.js]: Why do you think this should be removed?

File: lib/cast/int32.js (hunk 29 above)
Comment (vkarpov15) [reply lib/cast/int32.js]: I agree with using `_bsontype` rather than `instanceof`. Mongoose has an `isBsonType()` helper for this purpose.

The reason why we use `_bsontype` over `instanceof` is that it is common for Node.js apps to have multiple versions of bson floating around, and relying on `instanceof` can lead to hard-to-debug behavior when we get an Int32 instance from a different version of bson module

File: lib/schema/int32.js (hunk 21 above)
Comment (vkarpov15) [reply lib/schema/int32.js]: Yes, this function should check for integer.

The idea behind `_defaultCaster()` is that `_defaultCaster()` is the function Mongoose will use if casting is disabled using `cast: false`, so it should skip any type coercions (like converting strings or objects to int32).

File: lib/schema/int32.js (hunk 11 above)
Comment (vkarpov15) [reply lib/schema/int32.js:198]: I think this code is fine as written.

File: test/int32.test.js [hunk 30]
```diff
@@ -0,0 +1,512 @@
...
+
+    const doc = new Test({
+      myInt32: 13
+    });
+    assert.strictEqual(doc.myInt32, 13);
+    assert.equal(typeof doc.myInt32, 'number');
+  });
+
+  describe('supports the required property', function() {
+    it('when vaglue is null', async function() {
```
Comment (vkarpov15) [inline test/int32.test.js]: Minor typo value -> vaglue

File: test/int32.test.js [hunk 31]
```diff
@@ -0,0 +1,512 @@
...
+    it('casts from string', function() {
+      const schema = new Schema({
+        myInt: {
+          type: Schema.Types.Int32
+        }
+      });
+      const Test = mongoose.model('Test', schema);
+
+      const doc = new Test({
+        myInt: -42
```
Comment (vkarpov15) [inline test/int32.test.js]: If I'm reading this test correctly, myInt should be a string here

**Review by vkarpov15 — COMMENTED** [review]
I had a couple of minor comments, but overall this looks good, no major issues

File: index.js (hunk 27 above)
Comment (baileympearson) [reply index.js]: Not all schema types are exported as top-level exports, so I assumed that the exported schema types are exported for a reason.  Is that incorrect?

File: test/int32.test.js (hunk 31 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js]: Good catch!

File: test/int32.test.js (hunk 30 above)
Comment (aditi-khare-mongoDB) [reply test/int32.test.js]: Made the change, another good catch!

File: lib/cast/int32.js (hunk 29 above)
Comment (aditi-khare-mongoDB) [reply lib/cast/int32.js]: made the change!

File: lib/cast/int32.js (hunk 28 above)
Comment (aditi-khare-mongoDB) [reply lib/cast/int32.js]: Good catch, I didn't realize the checks here aren't necessary since the `valueOf` functions for `BSON.Double` and `BSON.Int32` are just returning `this.value`.

Comment (aditi-khare-mongoDB) [reply lib/cast/int32.js]: I've added in a unit test to flag the Long optimization!

File: lib/mongoose.js (hunk 26 above)
Comment (aditi-khare-mongoDB) [reply lib/mongoose.js]: Removed them - will add back in depending on discussion with Val

File: index.js (hunk 27 above)
Comment (aditi-khare-mongoDB) [reply index.js]: I've removed the schema type top level export for `Int32` but i can add it back in case this is incorrect behavior

File: lib/cast/int32.js [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+
+module.exports = function castInt32(val) {
+  if (val == null) {
+    return val;
+  }
+  if (val === '') {
+    return null;
+  }
+
+  const coercedVal = val instanceof BSON.Long ? val.toNumber() : Number(val);
```
Comment (nbbeeken) [inline lib/cast/int32.js]: https://github.com/mongodb-js/mongoose/pull/1#discussion_r1847157308

Shouldn't be using instanceof here per convo ^

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: lib/cast/int32.js (hunk 32 above)
Comment (aditi-khare-mongoDB) [reply lib/cast/int32.js]: My bad! Fixed now!

Comment (nbbeeken) [reply lib/cast/int32.js]: Per Val's comment, I think it is recommended to use the `isBsonType()` helper in mongoose

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: lib/cast/int32.js (hunk 32 above)
Comment (aditi-khare-mongoDB) [reply lib/cast/int32.js]: Oh I see! I'll change it again.

**Review by nbbeeken — APPROVED** [review]

**Review by baileympearson — COMMENTED** [review]
I can't approve because I opened the PR but LGTM
